### ISSUE TITLE
Mulitple RxMongoDrivers created for snapshot, journal, and readjrounal

### ISSUE THREAD

**Issue by mholtze** [scullxbones/akka-persistence-mongo#91]
The code below causes a different ConfiguredExtension to be created for each of snapshot, journal, readjournal, etc. The config that is used in the configuredExtensions map is different for each. This has side effects; for example in ReactiveMongo there would be a new ActorSystem created for each of journal, snapshot, ...

```
trait MongoPersistenceExtension extends Extension {
  def apply(config: Config): ConfiguredExtension = {
    configuredExtensions.putIfAbsent(config, configured(config))
    configuredExtensions.get(config).get
  }
  ...
}
```

**Comment by scullxbones** [comment]
Right - this is by design (config-per-plugin), working within the box that the akka-persistence plugin api provides.  I do see how that could cause lots of extra resource usage since RxMongo no longer requires an `ActorSystem` but instead creates one per `MongoDriver`.

So is this ticket a request to share a single `MongoDriver` per JVM?  I think that's fairly straightforward to do.

How is this issue affecting your project?  Want to make sure i'm working on the right thing.

**Comment by mholtze** [comment]
Thanks for your quick reply.

A little background. I discovered this issue creating a library  [bitbucket.org/mholtze/eventsource](https://bitbucket.org/mholtze/eventsource) on top of Akka Persistence to simplify bootstrapping event sourced programs. This library actually uses a fork of your akka-persistence-mongo project. My fork adds extra functionality including the ability to replay all events in the order they where created and handle serialization/deserialization to BSON without losing the message type. The fork is available at [github.com/mholtze/akka-persistence-mongo](https://github.com/mholtze/akka-persistence-mongo) and over the next couple months will get some usage/testing in live systems. I added the fix below to the fork for the multiple actor system issue.

To solve the creation of extra actor systems I modified RxMongoPersistenceExtension.scala and added the simple class below. RxMongoPersistenceExtension creates a single shared instance of MongoDriverProvider and provides it to each configured plugin.

    class MongoDriverProvider {
      lazy val driver = MongoDriver()
    }

**Comment by scullxbones** [comment]
Oh.  I didn't realize anyone had an actively maintained fork.  If you don't mind me asking, why not submit PRs?  I'd like to think i'm pretty open to them, and resolve them quickly, but if not - that's good feedback to have.

I think this plugin now does the "replay all events in natural order" via the `allEvents` read journal query, but there is no global sequence number.  I'd have to look deeper into the BSON work you've done to understand where the current version falls short.

**Comment by mholtze** [comment]
As for the PR's, I just had not reached a stable point until now. I was doing the modifications in conjunction with other development work. So, until now my changes were mainly experimental.

**For allEvents there are 3 things I needed to resolve:**

1. Without an explicit sort MongoDB makes no guarantee that documents will be retrieved in any particular order. They will be retrieved in what MongoDB calls "natural order", which is really just how the documents are ordered on disk. While this will often mirror insertion order, that is not assured. To assure that events can be replayed in the original insertion order I added a global sequence number to each.
2. allEvents does not have a way to specify a lower bound on the events to retrieve.
3. I needed live streaming of events that are saved without having to continually poll MongoDB. When I created my fork there was no support for live streaming. However, I did a quick check of the current master and it looks like live streaming may have been added.

**BSON Serialization**

The plugin only supported pass-through BSON for storing events as documents. I wanted events to have automatic serialization and deserialization from case classes to BSON and back using ReactiveMongo's BSON handler macro. So, I added support for event case classes with a companion object like that bellow.

    case class MyEvent(value: Int) extends BsonMessage

    object MyEvent extends RxMongoBsonHandler[MyEvent] {
      override implicit val bsonHandler = Macros.handler[MyEvent]
    }
	
This event will be automatically converted to a BSONDocument when it is saved and when the event is loaded it will be derserialized to MyEvent instead of remaining a BSONDocument.

**Comment by scullxbones** [comment]
That fix went in pretty cleanly.  I'm still working on some other issues, but this fix should be available in the next release.

**Comment by scullxbones** [comment]
This is fixed in release `1.1.9`.  Thanks for the report!

**Comment by scullxbones** [comment]
Issue not addressed, will be addressed by PR #94
